detector.go

```diff
@@ -1,118 +1,159 @@
-		if r.Metadata == nil {
-			r.Metadata = make(map[string]interface{})
+func (d *Detector) detectOrder(order buildpack.Order, done, next []buildpack.GroupBuildpack, optional bool, wg *sync.WaitGroup) ([]buildpack.GroupBuildpack, []BuildPlanEntry, error) {
+	ngroup := buildpack.Group{Group: next}
+	buildpackErr := false
+	for _, group := range order {
+		// FIXME: double-check slice safety here
+		found, plan, err := d.detectGroup(group.Append(ngroup), done, wg)
+		if err == ErrBuildpack {
+			buildpackErr = true
+		}
+		if err == ErrFailedDetection || err == ErrBuildpack {
+			wg = &sync.WaitGroup{}
+			continue
 		}
-		r.Metadata["version"] = r.Version
-		r.Version = ""
+		return found, plan, err
+	}
+	if optional {
+		return d.detectGroup(ngroup, done, wg)
 	}
-}
 
-func (r *Require) hasInconsistentVersions() bool {
-	if version, ok := r.Metadata["version"]; ok {
-		return r.Version != "" && r.Version != version
+	if buildpackErr {
+		return nil, nil, ErrBuildpack
 	}
-	return false
+	return nil, nil, ErrFailedDetection
 }
 
-func (r *Require) hasDoublySpecifiedVersions() bool {
-	if _, ok := r.Metadata["version"]; ok {
-		return r.Version != ""
+func (d *Detector) detectGroup(group buildpack.Group, done []buildpack.GroupBuildpack, wg *sync.WaitGroup) ([]buildpack.GroupBuildpack, []BuildPlanEntry, error) {
+	for i, bp := range group.Group {
+		key := bp.String()
+		if hasID(done, bp.ID) {
+			continue
+		}
+
+		bpi, err := d.Store.Lookup(bp.ID, bp.Version)
+		if err != nil {
+			return nil, nil, err
+		}
+		info := bpi.ConfigFile()
+
+		bp.API = info.API
+		bp.Homepage = info.Buildpack.Homepage
+		if info.Order != nil {
+			// TODO: double-check slice safety here
+			// FIXME: cyclical references lead to infinite recursion
+			return d.detectOrder(info.Order, done, group.Group[i+1:], bp.Optional, wg)
+		}
+		done = append(done, bp)
+		wg.Add(1)
+		go func() {
+			if _, ok := d.Runs.Load(key); !ok {
+				d.Runs.Store(key, bpi.Detect(&d.DetectConfig))
+			}
+			wg.Done()
+		}()
 	}
-	return false
-}
 
-func (r *Require) hasTopLevelVersions() bool {
-	return r.Version != ""
+	wg.Wait()
+
+	return d.Resolver.Resolve(done, d.Runs)
 }
 
-type Provide struct {
-	Name string `toml:"name"`
+func hasID(bps []buildpack.GroupBuildpack, id string) bool {
+	for _, bp := range bps {
+		if bp.ID == id {
+			return true
+		}
+	}
+	return false
 }
 
-type DetectConfig struct {
-	FullEnv       []string
-	ClearEnv      []string
-	AppDir        string
-	PlatformDir   string
-	BuildpacksDir string
-	Logger        Logger
-	runs          *sync.Map
+type DefaultResolver struct {
+	Logger Logger
 }
 
-func (c *DetectConfig) process(done []GroupBuildpack) ([]GroupBuildpack, []BuildPlanEntry, error) {
-	var runs []DetectRun
+// Resolve aggregates the detect output for a group of buildpacks and tries to resolve a build plan for the group.
+// If any required buildpack in the group failed detection or a build plan cannot be resolved, it returns an error.
```

Comment:
This is starting to look like two responsibilities - log aggregation and plan resolution. I'm not sure if separating them out will make the code more testable or easier to reason about...

Reply:
Interesting. I think that it might worth trying.

Reply:
We decided to punt on this for now.